### ISSUE TITLE
How to handle unique constraints, like unique user email address?

### ISSUE THREAD

**Issue by purefun** [dogmatiq/verity#422]
thx in advance.

**Comment by jmalloc** [comment]
Hi! This is a question that seems like it should be easy to answer but can get quite involved. 

I'll do my best to keep it relatively short and present a few options. The Dogma ecosystem is still in its infancy, and I'm aware that documentation is severely lacking so please feel free to ask follow up questions.

I've not found any of these approaches to be applicable in all use cases, selecting one depends heavily on the requirements of your application.

1) **Use the email address as the aggregate instance ID of the user** (not recommended)
  This is probably "easiest" way to enforce uniqueness that is tempting to try, but it causes too many headaches when you need to change the user's email address. I've included it for completeness/postery.

1) **Model the email address as a resource, instead of a property of a user**
  Give the email address its own [aggregate](https://github.com/dogmatiq/dogma#aggregate) (separate from the user), with commands like `AssignEmailAddressToUser`. This is the most "pure" way to implement such a constraint entirely within a Dogma application. 
    
1) **Detect conflicts using a process**
  Don't try to enforce uniqueness in the aggregate(s) at all, instead implement a [process](https://github.com/dogmatiq/dogma#process) to detect and resolve conflicts after they occur. This this _might_ not be a great solution for an email address, specifically if they're used as usernames, but can work well for other kinds of things we sometimes model as "unique", such as two users sharing a credit card.

1) **Use a projection to reject duplicates at validation time**
  This can/should be done in conjunction with any of the options above. Build a [projection](https://github.com/dogmatiq/dogma#projection) that can be queried to determine if an email address is in use already, and fail-fast at validation time. This can't _prevent_ duplicates outright, but in practice works well. This can also work well with a "conflict detecting process" as above.

1) **Consider not having unique email addresses, if possible**
  This sounds stupid, but I mention it just to point out that we often do things like this out of habit, and it's worth assessing the true nature of the requirement.

1) **Consider not using a Dogma application for user management**
  Sometimes the eventually-consistent nature of these systems is just not suitable for the problem we're trying to solve.

If you're like me, some of these ideas will make you feel a bit uneasy at first, and seem like a lot of work to do something that is very easy with a traditional RDBMS. The reason this is difficult is because the Dogma abstraction is designed to allow running a [Dogma application](https://github.com/dogmatiq/dogma#application) in a distributed manner, where it's not necessarily easy to check that a value is globally unique.

I hope this is helpful :)

**Comment by purefun** [comment]
Thank you for your thorough analysis. I'm so surprised. The second approach sounds great. Does it mean `EmailAddress` instance ID's value is the email address? If a user email changed, just record an event like `ReassignEmailAddress`.

Is `dogmatiq/verity` used for your personal or company real applications? What're applications look like?

**Comment by jmalloc** [comment]
> Thank you for your thorough analysis. I'm so surprised. The second approach sounds great. 

No worries, you're welcome :)

> Does it mean `EmailAddress` instance ID's value is the email address?  If a user email changed, just record an event like `ReassignEmailAddress`.

Yep, correct on both points (assuming `EmailAddress` is the name of an aggregate here).  I have prototyped a similar approach for usernames in the past. In that project I used separate events for `UsernameUnassigned`, `UsernameAssigned`  (note past-tense naming for events, btw).

> Is `dogmatiq/verity` used for your personal or company real applications? What're applications look like?

We are using it at our day job (the other members of the dogmatiq organisation work at the same company as I do). Unfortunately, I can't share any of the code from those projects. The best I can offer is the [dogma example app](https://github.com/dogmatiq/example), though I suspect you've already seen it.

FWIW, we're in the process of expanding that example to include a web-based UI and browser-facing API to demonstrate one way that everything can be wired together, but it's not quite ready yet.
